src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -29,7 +29,7 @@ public sealed partial class JsonSerializerOptions
 
         internal JsonSerializerContext? _context;
 
-        private Func<Type, JsonSerializerOptions, JsonTypeInfo>? _typeInfoCreationFunc = null!;
+        private static Func<Type, JsonSerializerOptions, JsonTypeInfo>? _typeInfoCreationFunc = null!;
```

Comment:
Are we sure this should be static? How does this interact when I have 2 Options instances, one from source gen and one not from source gen. If I try to serialize something from source gen before using the non-source gen options, it will throw an exception. But if the ordering gets changed, suddenly the same operation will stop throwing because the non-source gen options affected static state.

If this should really be static, the field should be named `s_typeInfoCreationFunc`.

Also, what is the value in saying this isn't nullable, but then we check it for null later in the code?

Reply:
What is the downside of not having an exception (i.e. having things work) in the case where you used source-gen overloads first? Chances are you shouldn't even hit the code-path where we could invoke this func. If all root serializable types are declared with `JsonSerializableAttribute`, along with any types that could be serialized polymorphically, then you'll never hit the code-path where we try to create the metadata using reflection.

One downside is that folks might not configure their source-gen options correctly and wouldn't know because "things just work". Is this the problem we are avoiding by making it an instance field? Folks that care about size and not doing run-time reflection would likely profile their apps to verify.

Making the field static would mean avoiding an extra alloc for each `options` instance. cc @steveharter in case that was the perspective from https://github.com/dotnet/runtime/pull/51300#discussion_r614133566.

---

> Also, what is the value in saying this isn't nullable, but then we check it for null later in the code?

It is nullable - I'll remove the ` = null!` as that is not needed.

Reply:
> What is the downside of not having an exception (i.e. having things work) in the case where you used source-gen overloads first?

In general, having static mutable state is bad, for many reasons.

Here's the case I'm concerned about:

Working code:

```C#
// 1. Code uses non-source gen serialization at the beginning of the app
var options = new JsonSerializationOptions();
object someObject = ...;
JsonSerializer.Serialize(someObject, options);

// The above code "initializes" the static variable, causing the below code to "work"

// 2. Somewhere else in the app, code is using the source gen APIs, but they don't have `JsonSerializableAttribute` for a type. The code in (1) above executed first, so this code executes successfully:

JsonSerializer.Serialize(stronglyTypeObject, MyJsonContext);
```

Sometime later, the app changes and no longer needs to do `(1)` above. So that code gets deleted.

No longer working code:

```C#
// This now throws somewhere completely unrelated to removing (1) above. Note that (1) and (2) may be in unrelated referenced libraries. 
JsonSerializer.Serialize(stronglyTypeObject, MyJsonContext);
```

Reply:
Yes that makes sense. One could still expect that people using the source-gen pattern do the right thing, but it will be hard to correct mistakes if they span across different assemblies.